compiler/src/iree/compiler/Dialect/Stream/Transforms/MemoizeChannels.cpp

```diff
@@ -0,0 +1,115 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <utility>
+
+#include "iree/compiler/Dialect/Stream/IR/StreamOps.h"
+#include "iree/compiler/Dialect/Stream/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Stream/Transforms/Passes.h"
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "llvm/ADT/StringSet.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Stream {
+
+// NOTE: this implementation is just for a single active device. As we start to
+// support multiple devices we'll need to change this to be per-device.
```

Comment:
I'm a bit worried about how many places assume single device. Do you think we'll be able to incrementally support multiple devices without some of these more advanced features (collectives in this case)?

Reply:
There's just a few passes at this point, but yeah, we'll need to fix them. My plan is to have one convention we use like `emitForAllDevices(...callback...)` and that these passes can all be updated for that - that helper could then emit the device selection logic and duplication if there are multiple devices live at a time.